Added cluster distribution extension

jroper opened the pull request:
This will be used by sharded read side processing to distribute the aggregate tags over the cluster.

It could be a useful utility to include in akka-cluster-sharding.

A few things that could be changed that I didn't do:

* We could ensure that the entity actors don't need to handle the `EnsureActive` message by creating the user supplied props in a our own entity actor.
* We could have the scheduled ping managed by one system actor instead of one per cluster distribution.

jroper commented:
@patriknw This is ready for review.

patriknw commented on persistence/src/main/resources/reference.conf:
```diff
@@ -213,3 +213,10 @@ akka.actor {
 }
 
+# Cluster distribution settings
+akka.cluster.distribution {
```
I'd recommend against using the `akka` namespace for things that don't live in Akka

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+  }
+}
+
+object ClusterDistribution extends ExtensionId[ClusterDistribution] with ExtensionIdProvider {
```
also override `get`, otherwise it will not have a proper java api

`override def get(system: ActorSystem): ClusterDistribution = super.get(system)`

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+  /**
+   * Sent to each entity in a cluster distribution to ensure it's active.
+   */
+  case class EnsureActive(entityId: EntityId)
```
final

patriknw replied:
Use `akka.actor.Identify` instead. It's a built in message that all actors reply to, and it will not be delivered to user's `receive`. Then you will get `ActorIdentity` messages back so you should receive and ignore them.

jroper replied:
Won't `Identify` get handled by the shard region actor then though?  ie, it will be responded to directly by the shard region actor, rather than delivered to its receive, and so won't trigger the creation of the shard?

patriknw replied:
you're right, my bad

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+  def start(
+    typeName:    String,
+    entityProps: Props,
+    entityIds:   Seq[EntityId],
```
use `immutable.Seq`

patriknw replied:
isn't it a `Set` by the way

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+      case msg @ EnsureActive(entityId) => (entityId, msg)
+    }
+    val extractShardId: ShardRegion.ExtractShardId = {
+      case EnsureActive(entityId) => entityId
```
ok, one shard per entity, probably alright since it shouldn't be too many partitions
might be good with an upper limit

jroper replied:
The intention was that the entities are actually shards - afterall it's a static list and so small.  That said, we could have a shard function that turns an entity ID into a shard.

One question that I have, why isn't ExtractShardId a function of entity id -> shard id? Since the entity id has already been extracted out by ExtractEntityId?  My original intention was to allow the user to supply a custom extract entity id function which would mean they could then send their own messages to the entities as well, since I could combine the EnsureActive handling with orElse. But since extract shard id works off actual messages, and not entity IDs, they'd also have to supply a custom extract shard id function, but that would be a problem since the hash function of entity id to shard id must be consistent, we wouldn't be able to inject our own EnsureActive message handling to the extract shard id function, so the user would have to handle it in their own extract shard id.

I guess we can use the user defined extract entity id function to produce the extract shard id function, but it just seems odd to me the way it is, extract shard id could end up sending messages of the same entity id to different shards, is there ever a use case for doing that?

patriknw replied:
> why isn't ExtractShardId a function of entity id -> shard id

Yeah, that would be enough for almost all cases. The idea was that it should be possible to derive the shard id from more properties than only the entity id.

I see no reason for custom messages here, use `Identify`.

It wouldn't hurt to cap the number of shards with something like `entityId % 1000`. Doesn't have to be configurable.

jroper replied:
`ShardId` is a `String` not an `int`, so would need to be hashed, it could hurt if you have a small number of entity ids and two of them hash to the same value.  I'll cap iff the number of entity IDs exceeds 1000.

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+}
+
+private[cluster] object EnsureActiveActor {
+  case object Tick
```
final

patriknw replied:
`with DeadLetterSuppression`

jroper replied:
Didn't know there was a DeadLetterSuppression, are there any rules around when to use/not use it?

patriknw replied:
It's only for reducing logging noise. Good to use it on scheduled messages since there is a chance they arrive after the actor is stopped.

jroper replied:
Also, IntelliJ tells me, and I just confirmed with scalac and javap, that `final` is redundant on an object.

patriknw replied:
sorry, not needed, I triggered on `case`

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,128 @@
+      entityIds.foreach { entityId =>
+        shardRegion ! EnsureActive(entityId)
+      }
+    case Terminated(subject) if subject == shardRegion =>
```
case Terminated(`shardRegion`)

with backticks

patriknw commented:
LGTM, I would happily accept this as an addition to Akka

ktoso commented:
I like it, would indeed be nice addition to Akka itself 👍

jroper commented:
I've addressed all review comments, except:

* Didn't switch to `Identify` because I'm pretty sure it will be handled by Akka too early to achieve its purpose.
* Didn't make the case object final since objects are implied final.

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,142 @@
+  /**
+   * Sent to each entity in a cluster distribution to ensure it's active.
+   */
+  final case class EnsureActive(entityId: EntityId)
```
You should add a proper serializer for this message to avoid Java serialization. 

If in Lagom it can be added to https://github.com/lagom/lagom/blob/master/persistence/src/main/scala/com/lightbend/lagom/internal/persistence/protobuf/PersistenceMessageSerializer.scala

If in Akka it can be added in https://github.com/akka/akka/blob/master/akka-cluster-sharding/src/main/scala/akka/cluster/sharding/protobuf/ClusterShardingMessageSerializer.scala

patriknw commented on persistence/src/main/scala/com/lightbend/lagom/internal/persistence/cluster/ClusterDistribution.scala:
```diff
@@ -0,0 +1,142 @@
+    val extractEntityId: ShardRegion.ExtractEntityId = {
+      case msg @ EnsureActive(entityId) => (entityId, msg)
+    }
+    val extractShardId: ShardRegion.ExtractShardId = if (entityIds.size > MaxShards) {
```
good

jroper commented:
Added serializers for EnsureActive, also I made a small change, it will only start the sharding and the ensure active actor on roles that it's configured to run on.

patriknw commented:
LGTM
